### DISCUSSION TITLE
Will NNI support a customized deep learning framework developed under OneAPI?

### DISCUSSION THREAD

**Discussion by goooxu** [microsoft/nni#3823] [Ideas]
**Background**

> oneAPI is a cross-industry, open, standards-based unified programming model that delivers a common developer experience across accelerator architectures—for faster application performance, more productivity, and greater innovation. The oneAPI industry initiative encourages collaboration on the oneAPI specification and compatible oneAPI implementations across the ecosystem.

We known that Intel released oneAPI Toolkits recently, which provides a unified, standards-based programming model, our ML training system is just developed under oneAPI.
The NNI toolkit doesn't only support Tensorflow or Pytorch, am I right? So, how to tune a customized native training system (e.g. written in DPC++ and oneAPI) via NNI. Because the customized training system has an obvious advantage in performance, so it uses a lot in industry.

 Hope for reply.

**Comment by scarlett2018** [comment]
@goooxu - Yes, NNI supports DL training in Tensorflow and Pytorch, as long as the program is written in Python. If the native training with OneAPI you mentioned is written in Python, theoretically, NNI should able to support it.

On the other hands, if I understood correctly, the OneAPI is more a unique API layer for connect, utilize and accelerate the interact between Applications and Hardware. so, are you more suggesting a OneAPITrainingService here? As the team is not familiar with OneAPI, would you like to take over this implementation? we have very detailed document about how to [implement a customized training service](https://github.com/microsoft/nni/blob/master/docs/en_US/TrainingService/HowToImplementTrainingService.md), and will provide review and supports if there are any questions around the implementation.

**Comment by goooxu** [comment]
@scarlett2018 Yes, the OneAPI is a C++ layer for the interaction between Application and Hardware, so our customized machine learning framework is written in C++.
I checked the documents of NNI but didn't find anywhere mentions of how to support a C ++ training framework (similar to Caffe), will NNI support it in the future?
